### PR TITLE
18f-2811 Fix issue around expired account invitation links

### PR DESCRIPTION
**Pull request by rememberlenny**
# What

https://favro.com/card/1e11108a2da81e3bd7153a7a/18F-2811

### CODE REVIEW

File: uaaextras/templates/email/body.html [hunk 1]
```diff
@@ -25,7 +25,7 @@
                 To log into cloud.gov, which requires multi-factor authentication, first install a mobile application that generates time-based one-time passwords.  We recommend installing either <a href="https://support.google.com/accounts/answer/1066447?hl=en">Google Authenticator</a> or <a href="https://www.authy.com/app/mobile">Authy</a> on your mobile device. In your next step, the cloud.gov registration process will give you instructions for using it.
             </li>
             <li>
-                <a href="{{invite.inviteLink}}">Accept your invite</a> to continue the registration process.
+                <a href="{{verification_url}}">Accept your invite</a> to continue the registration process.
```
Comment (rememberlenny) [inline uaaextras/templates/email/body.html]: @cnelson Im not sure if I am able to do this switch. Your input would be great.

File: uaaextras/webapp.py [hunk 2]
```diff
@@ -38,6 +38,7 @@
     'IDP_PROVIDER_URL': 'https://idp.bosh-lite.com',
     'PW_EXPIRES_DAYS': 90,
     'PW_EXPIRATION_WARN_DAYS': 10,
+    'UAA_INVITE_EXPIRATION_IN_SECONDS': timedelta(days=7),
```
Comment (rememberlenny) [inline uaaextras/webapp.py]: @rogeruiz Where should I get the time for expiration here?

Comment (rogeruiz) [inline uaaextras/webapp.py]: @rememberlenny https://github.com/18F/cg-deploy-cf/blob/b343574853e31a4bc6ff13c00d836b0013ca184f/cf-properties.yml#L318-L321

File: uaaextras/webapp.py [hunk 3]
```diff
@@ -481,6 +482,37 @@ def signup():
 
         return render_template('error/internal.html'), 500
 
+    @app.route('/redeem-invite', methods=['GET'])
+    def redeem_invite():
+
+        if request.method == 'GET':
```
Comment (jmcarp) [inline uaaextras/webapp.py]: Since this route only responds to GET, can we drop this check?

Comment (rememberlenny) [inline uaaextras/webapp.py]: Yep!

File: uaaextras/templates/email/body.html (hunk 1 above)
Comment (cnelson) [inline uaaextras/templates/email/body.html]: LGTM

File: uaaextras/webapp.py [hunk 4]
```diff
@@ -481,6 +482,37 @@ def signup():
...
+            if r:
+                # Check if Redis key was previously requested, otherwise load the UAA invite
+                uaaInviteLink = r.get(verification_code)
+                if uaaInviteLink is None:
+                    logging.info('UAA invite link is expired. {0}'.format(verification_code))
+                    return render_template('error/token_validation.html'), 401
+
+                else:
+                    logging.info('Accessed UAA invite link. {0}'.format(verification_code))
+                    r.delete(verification_code)
```
Comment (cnelson) [inline uaaextras/webapp.py]: I think if we delete the code here, we are just recreating the problem UAA has.  We need the user to take an action before clearing this key or we'll still have links invalidated by mail scanners

File: uaaextras/webapp.py [hunk 5]
```diff
@@ -481,6 +482,37 @@ def signup():
 
         return render_template('error/internal.html'), 500
 
+    @app.route('/redeem-invite', methods=['GET'])
+    def redeem_invite():
+
+        if request.method == 'GET':
+            # Store the validation token to check for UAA invite link
+            verification_code = request.args.get('validation_token')
```
Comment (cnelson) [inline uaaextras/webapp.py]: Probably cleaner to move this below the check / error if it doesn't exist

Comment (rememberlenny) [inline uaaextras/webapp.py]: Fixed

File: uaaextras/webapp.py (hunk 4 above)
Comment (rememberlenny) [inline uaaextras/webapp.py]: I took out the delete and placed the link behind a button.

Comment (rememberlenny) [inline uaaextras/webapp.py]: @cnelson This should be good now! I think you might need to "approve" the review.

**Pull request merged**